File: n8n-discord-bot/gateway-bridge/src/observability.js
// Sentry error monitoring for the Discord -> n8n gateway bridge.
//
// Sentry init is intentionally optional — when SENTRY_DSN is unset (local
// dev, CI lint, anyone forking the bridge) every helper here becomes a
// cheap no-op so the process still starts and runs identically to before.
//
// Operationally, set the DSN as a Fly.io secret:
//
//     flyctl secrets set SENTRY_DSN='https://<key>@oXXX.ingest.de.sentry.io/YYY'
//
// Optional env vars:
//
//     SENTRY_ENVIRONMENT          default: production
//     SENTRY_RELEASE              default: unset (Sentry auto-detects git SHA)
//     SENTRY_TRACES_SAMPLE_RATE   default: 0.0 (errors only — cheaper)

let _initialized = false;
let _sentry = null;

export async function initSentry() {
  if (_initialized) return true;

  const dsn = (process.env.SENTRY_DSN || '').trim();
  if (!dsn) {
    console.log('[sentry] SENTRY_DSN unset — Sentry disabled');
    return false;
  }

  try {
    _sentry = await import('@sentry/node');
  } catch (err) {
    console.warn('[sentry] @sentry/node not installed — Sentry disabled');
    return false;
  }

  let sampleRate = 0.0;
  const raw = process.env.SENTRY_TRACES_SAMPLE_RATE;
  if (raw) {
    const parsed = Number(raw);
    if (Number.isFinite(parsed)) sampleRate = parsed;
  }

  try {
    _sentry.init({
      dsn,
      environment: process.env.SENTRY_ENVIRONMENT || 'production',
      release: process.env.SENTRY_RELEASE || undefined,
      tracesSampleRate: sampleRate,
      // Discord bot tokens, bridge HMAC secrets, and n8n webhook URLs must
      // never leave the process. ``sendDefaultPii`` defaults to false
      // already; pin it explicitly so a future SDK default flip can't
      // surprise us.
      sendDefaultPii: false,
      maxBreadcrumbs: 50,
    });
  } catch (err) {
    // @sentry/node.init is extremely defensive today, but observability is
    // optional and must NEVER take the bridge down. Log and fall through
    // to the disabled state.
    console.error('[sentry] init failed — Sentry disabled:', err?.message);
    return false;
  }

  _initialized = true;
  console.log(
    `[sentry] initialized (env=${process.env.SENTRY_ENVIRONMENT || 'production'})`,
  );
  return true;
}

// Capture an exception with optional extra context. No-op if disabled.
//
// Extras are scoped to this single call via ``withScope()`` so they
// cannot leak into unrelated Sentry events fired later (e.g. another
// gateway event raised on the same client instance).
export function captureException(err, extra = {}) {
  if (!_initialized || !_sentry) return;
  try {
    _sentry.withScope((scope) => {
      for (const [key, value] of Object.entries(extra)) {
        scope.setExtra(key, value);
      }
      _sentry.captureException(err);
    });
  } catch {
    // Last-resort guard: a broken Sentry transport must never crash the
    // caller. Swallow and move on.
  }
}

// Wire up Node's global crash handlers so an uncaught throw on a
// background promise or in an event listener doesn't disappear into
// stderr. discord.js intentionally swallows event-listener errors in
// some cases, so without this they would be invisible in production.
export function installGlobalErrorHandlers() {
  if (!_initialized) return;
  process.on('unhandledRejection', (reason) => {
    const err = reason instanceof Error ? reason : new Error(String(reason));
    captureException(err, { source: 'unhandledRejection' });
  });
  process.on('uncaughtException', (err) => {
    captureException(err, { source: 'uncaughtException' });
    // Match Node's default behavior: log + let the process die. Don't
    // resurrect a crashed process — Fly's machine restart is the right
    // recovery path.
    console.error('[fatal] uncaughtException', err);
  });
}

File: boon-community-bot/bot/observability.py
"""Sentry error/performance monitoring for the BOON Community Bot.

Sentry init is intentionally optional — when ``SENTRY_DSN`` is unset
(local dev, CI lint, anyone forking the bot) all helpers in this module
become cheap no-ops, so the bot still runs identically to before.

The slash-command error handler in ``main.py`` passes Discord-specific
context (command name, guild ID, user ID) as ``extra`` kwargs to
``capture_exception``, which scopes them to that single Sentry event
via ``sentry_sdk.new_scope()``. This avoids the classic ``Hub``-scope
leak where tags from one error report bleed into the next.

Operationally, set the DSN as a Fly.io secret:

    flyctl secrets set SENTRY_DSN='https://<key>@oXXX.ingest.de.sentry.io/YYY'

Optional env vars:

    SENTRY_ENVIRONMENT   default: production
    SENTRY_RELEASE       default: unset (lets Sentry auto-detect git SHA)
    SENTRY_TRACES_SAMPLE_RATE  default: 0.0 (errors only — cheaper)
"""

from __future__ import annotations

import logging
import os
from typing import Any

log = logging.getLogger("boon-bot.observability")

_INITIALIZED = False


def init_sentry() -> bool:
    """Initialize Sentry if ``SENTRY_DSN`` is set. Returns True on success."""
    global _INITIALIZED
    if _INITIALIZED:
        return True

    dsn = os.environ.get("SENTRY_DSN", "").strip()
    if not dsn:
        log.info("SENTRY_DSN unset — Sentry disabled")
        return False

    try:
        import sentry_sdk
        from sentry_sdk.integrations.aiohttp import AioHttpIntegration
        from sentry_sdk.integrations.logging import LoggingIntegration
    except ImportError:
        log.warning("sentry-sdk not installed — Sentry disabled")
        return False

    try:
        sample_rate = float(os.environ.get("SENTRY_TRACES_SAMPLE_RATE", "0.0"))
    except ValueError:
        sample_rate = 0.0

    sentry_sdk.init(
        dsn=dsn,
        environment=os.environ.get("SENTRY_ENVIRONMENT", "production"),
        release=os.environ.get("SENTRY_RELEASE") or None,
        traces_sample_rate=sample_rate,
        # INFO+ logs become breadcrumbs. ``event_level=None`` suppresses
        # automatic event creation from ERROR-level log records, so the
        # only Sentry events are the ones the bot explicitly captures
        # via ``capture_exception`` — avoids the double-report path
        # where ``log.exception(...) + capture_exception(...)`` sitting
        # next to each other in main.py would otherwise generate two
        # separate Sentry events (DedupeIntegration is best-effort).
        integrations=[
            AioHttpIntegration(),
            LoggingIntegration(level=logging.INFO, event_level=None),
        ],
        # Bot tokens, github PATs, and discord webhook signing secrets must
        # never leave the process. ``send_default_pii`` defaults to False
        # already; pin it explicitly so a future SDK default flip can't
        # surprise us.
        send_default_pii=False,
        max_breadcrumbs=50,
    )
    _INITIALIZED = True
    log.info("Sentry initialized (env=%s)",
             os.environ.get("SENTRY_ENVIRONMENT", "production"))
    return True


def capture_exception(exc: BaseException, **extra: Any) -> None:
    """Capture an exception with optional extra context. No-op if disabled.

    Extras are scoped to this single call via ``sentry_sdk.new_scope()``
    so they cannot leak into unrelated Sentry events fired later from
    the same task (or from child tasks that inherit the contextvar).
    """
    if not _INITIALIZED:
        return
    try:
        import sentry_sdk
    except ImportError:
        return
    with sentry_sdk.new_scope() as scope:
        for key, value in extra.items():
            scope.set_extra(key, value)
        sentry_sdk.capture_exception(exc)
